### PR TITLE
Session Lun 30/11

### PR DESCRIPTION
**Pull request by pleasedontbelong**

### CODE REVIEW

File: modules/api.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import requests
+import json
+
+
+class OpenFactFood:
+
+
+    def __init__(self, nb_cat, nb_prod):
+
+        self.nb_cat = nb_cat
```
Comment (pleasedontbelong) [inline modules/api.py:20]: nb_categories

File: modules/api.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        self.get_top_categories()
+        self.get_products()
+
+    def log(self, msg):
+        print(msg)
+
+    def get_top_categories(self):
+        self.top_cat = [None for e in range(self.nb_cat)]
+
+        r = requests.get('https://fr.openfoodfacts.org/categories.json')
```
Comment (pleasedontbelong) [inline modules/api.py]: move this to a function on another module "requester"

File: modules/api.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                    self.top_cat.insert(i, category)
+                    self.top_cat.pop(len(self.top_cat) - 1)
+                    break
+
+
+    def get_products(self):
+        if self.err is not None:
+            return
+
+        for j in range(len(self.top_cat)):
```
Comment (pleasedontbelong) [inline modules/api.py]: `for category in self.top_cat`

File: modules/database.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    def create_table(self, name, *col, **options):
+        values = ()
+
+        sql = "CREATE TABLE IF NOT EXIST {} ("
+        values.append(name)
+        for i, x in enumerate(col):
+            if i != 0:
+                sql += ', '
+            sql += '{}'
+            values.append(x)
```
Comment (pleasedontbelong) [inline modules/database.py]: use `join()` maybe

File: modules/database.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        sql = "INSERT INTO {} ("
+        values.append(table)
+        for i, val_name in enumerate(values_order):
+            if i != 0:
+                sql += ', '
+            sql += '{}'
+            values.append(val_name)
+        sql += ") VALUES "
+
+        for j, line_vals in enumerate(values):
```
Comment (pleasedontbelong) [inline modules/database.py]: use join

File: modules/api.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    def __init__(self, nb_cat, nb_prod):
+
+        self.nb_cat = nb_cat
+        self.categories = []
+        self.nb_prod = nb_prod
+        self.products = []
+        self.get_categoriesegories()
+
+
+    def get_categoriesegories(self):
```
Comment (pleasedontbelong) [inline modules/api.py]: `get_categories`

File: modules/database.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        sql = "CREATE TABLE {} (".format(name)
+        sql += (', '.join(col))
+        sql += (')')
+
+        for arg in options:
+            sql += (' {}={}'.format(arg.upper(), options[arg].upper()))
+
+        return self.execute(sql)
+
+    def insert(self, table, values_names, values):  # values_names --> list, values --> list of dicts
```
Comment (pleasedontbelong) [inline modules/database.py]: comments on methods/functions should be inside the function name and using triple quotes:
```
def foo():
    """
    This is a comment
    """"
```

File: modules/api.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            }
+
+            r = requester('https://world.openfoodfacts.org/cgi/search.pl', params=payload)
+
+            self.products.append(r['products'])
+
+
+def requester(self, url, **kwargs):
+    if ('params' in kwargs):
+        r = requests.get(url, kwargs['params'])
```
Comment (pleasedontbelong) [inline modules/api.py:87]: I think you can just do:
```
r = requests.get(url, **kwargs)
```
so you don't need to pass a `params` parameter

File: modules/api.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    else:
+        r = requests.get(url)
+
+    if r.ok:
+        return json.loads(r.text)
+    else:
+        logging.error('Une erreur s\'est produite lors de la récupération des données')
+        logging.error('Code erreur HTTP : ' + r.status_code)
+        return None
```
Comment (pleasedontbelong) [inline modules/api.py:97]: I think we should raise an exception and catch that same exception at the top level

File: modules/database.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            if err.errno == errorcode.ER_TABLE_EXISTS_ERROR:
+                logging.error("Table already exists.")
+            elif err.errno == errorcode.ER_NO_SUCH_TABLE_ERROR:
+                logging.error("Table doesn't exist")
+            elif err.errno == errorcode.ER_WRONG_COLUMN_NAME_ERROR:
+                logging.error("Column doesn't exist")
+            else:
+                logging.error(err.msg)
+            logging.error(sql.format(values))
+            return 1
```
Comment (pleasedontbelong) [inline modules/database.py:64]: I'm not sure if you need to return something. maybe just re-raise the error

File: modules/database.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        else:
+            logging.info("Succeeded")
+            logging.info(sql.format(values))
+            return 0
+
+    def create_table(self, name, *col, **options):
+
+        sql = "CREATE TABLE {} (".format(name)
```
Comment (pleasedontbelong) [inline modules/database.py]: if you use formated strings, this could look cleaner

File: modules/api.py (hunk 6 above)
Comment (titou386) [reply modules/api.py]: Yes, my mistake ...
i pushed too quickly.

File: modules/api.py (hunk 8 above)
Comment (titou386) [reply modules/api.py:87]: yes, an another mistake ...

**Pull request merged**